domain/schema/model/sql/0019-application.sql

```diff
@@ -200,6 +200,42 @@ CREATE TABLE application_status (
     REFERENCES workload_status_value (id)
 );
 
+CREATE TABLE device_constraint (
+    application_uuid TEXT NOT NULL,
+    name TEXT NOT NULL,
+    type TEXT NOT NULL,
+    count INT,
+    attribute_uuid TEXT,
+    CONSTRAINT fk_device_constraint_application
+    FOREIGN KEY (application_uuid)
+    REFERENCES application (uuid),
+    CONSTRAINT fk_device_constraint_attribute
+    FOREIGN KEY (attribute_uuid)
+    REFERENCES device_constraint_attribute (uuid),
+    PRIMARY KEY (application_uuid, name, attribute_uuid)
+);
+
+CREATE TABLE device_constraint_attribute (
+    uuid TEXT NOT NULL PRIMARY KEY,
```

Comment:
This looks the wrong way around.

This allows a device constraint for one application to reference an attribute pointing at another.

It seems to me that there should be a UUID on device_constraint (a key value pair is not a domain entity) and device_constraint_attribute should have an FK reference to it.

Reply:
Yes that's better. I'd like to leave the `application_uuid` though because it's handy for deletion by reference, is that OK?

Reply:
No, because it says a thing about the data-model. 

If your primary key is the pair, then you have expressed that the same key/value could exist against multiple applications with the same device constraint identifier or vice versa.

The data-model describes the domain. The domain doesn't change to accommodate implementation; it's the other way around.

Approved contingent upon removing the app ID from the attribute table.